### ISSUE TITLE
Warn on redundant 'as Foo' statements

### ISSUE THREAD

**Issue by Wilfred** [rust-lang/rust#27445]
```rust
fn main() {
    let x: i8 = 42;
    let y: i8 = x as i8;
    println!("y: {}", y);
}
```

This doesn't generate any warnings from rustc, even though the `as i8` is redundant here. I bump into this occasionally when refactoring, so it would be nice if rustc could help find these sitautions.

**Comment by sfackler** [comment]
The `trivial-numeric-casts` lint will do this if you enable it.

**Comment by Aatch** [comment]
I guess the question is if we should make it a warning. The issue *there* being in macros, but you could always just turn them off in those cases?

**Comment by sfackler** [comment]
Redundant casts are also common when dealing with platform specific typedefs like in liblibc.

**Comment by steveklabnik** [comment]
Yes, we have a lint for this, and you can enable it. Turning it into a warn by defualt would need an RFC, I would imagine, as it's quite possible that it becomes a breaking change for a lot of code.
